app/controllers/callbacks_controller.rb

```diff
@@ -2,80 +2,90 @@
+    discord_id = discord_info['id']
 
     error_codes = []
+    identity_rows = []
 
     user_connections = get_user_connections(token)
-    user_connections.each do |connection|
-      Rails.logger.info connection
-      next unless connection['type'] == 'steam'
-
-      is_verified = connection['verified']
-      error_codes.push('steam-not-verified') unless is_verified
-      break unless is_verified
 
-      steam_id = connection['id']
-
-      steam_id_used = Identity.find_by(platform: 'steam', identifier: steam_id)
-      discord_id_used = Identity.find_by(platform: 'discord', identifier: discord_id)
-
-      error_codes.push('used-steam-id') if steam_id_used
-      error_codes.push('used-discord-id') if discord_id_used
+    # Array of hashes to be used in a where chain
+    connection_pairs = user_connections.map do |c|
+      { identifier: c['id'], platform: c['type'] }
+    end
 
-      break if steam_id_used || discord_id_used
+    # Generate a query to find any Identity with the
+    # specific pairs of identifiers and platforms
+    # TODO: Extract to model?
+    found_identities = connection_pairs.inject(Identity.none) do |memo, pair|
+      memo.or(Identity.where(pair))
+    end.pluck(:identifier, :user_id).to_h
 
-      new_user = User.create
-      Identity.create(user_id: new_user.id, platform: 'steam', identifier: steam_id)
-      Identity.create(user_id: new_user.id, platform: 'discord', identifier: discord_id)
+    ## TODO: DRY these two conditions
+    if connection_pairs.select { |c| c[:platform] == 'steam' }.empty?
+      error_codes << 'discord-missing-steam'
+    end
 
-      return redirect_to SUCCESS_URL
+    if connection_pairs.select { |c| c[:platform] == 'discord' }.empty?
+      identity_rows << Identity.new(platform: 'discord', identifier: discord_id)
     end
 
-    error_codes.push('discord-missing-steam') if error_codes.empty?
+    user_connections.each do |connection|
+      platform = connection['type']
+      identifier = connection['id']
+      is_verified = connection['verified']
 
-    error_codes = error_codes.join(',')
-    redirect_to "#{FAILURE_URL}?errors=#{error_codes}"
-  end
+      if is_verified == false
+        error_codes << 'steam-not-verified' if platform == 'steam'
+        next
+      end
 
-  def receive_cfc_bot_callback
-    steam_id = params['steam_id']
-    discord_id = params['discord_id']
+      user_id = found_identities[identifier]
 
-    error_codes = []
+      if user_id.nil?
+        identity_rows << Identity.new(platform: platform, identifier: identifier)
+      end
+    end
 
-    steam_id_used = Identity.find_by(platform: 'steam', identifier: steam_id)
-    discord_id_used = Identity.find_by(platform: 'discord', identifier: discord_id)
+    # We have to handle the situation where they may have multiple Users created
+    users = User.includes(:identities)
+                .where(id: found_identities.values)
+                .order('created_at DESC')
```

Comment:
`created_at: 'DESC'` desc probably has a valid symbol form too but doing it this way should allow activerecord to understand/override an already existing order by clause if they conflict (although I think in many cases it just overrides them anyway)